crates/state-asm/src/effects.rs

```diff
@@ -0,0 +1,98 @@
+use crate::StateRead as StateReadOp;
+use bitflags::bitflags;
+use essential_constraint_asm::{Access, Constraint as ConstraintOp};
+
+/// Flags to indicate the effects of state read operations.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct Effects(u8);
+
+bitflags! {
+    impl Effects: u8 {
+        /// Flag for [´StateReadOp::KeyRange´]
+        const Range = 1 << 0;
+        /// Flag for [´StateReadOp::KeyRangeExtern´]
+        const RangeExtern = 1 << 1;
+        /// Flag for [´StateReadOp::Constraint::Access::ThisAddress´]
+        const ThisAddress = 1 << 2;
+        /// Flag for [´StateReadOp::Constraint::Access::ThisContractAddress´]
+        const ThisContractAddress = 1 << 3;
+    }
+}
+
+/// Determine effects of the given state read program.
+pub fn determine_effects(ops: &[StateReadOp]) -> Effects {
```

Comment:
Could also be
```suggestion
pub fn effects(ops: &[StateReadOp]) -> Effects {
```
or maybe
```suggestion
pub fn analyze(ops: &[StateReadOp]) -> Effects {
```
so that you could use it like `asm::effects::analyze(&ops)`? :thinking: 

But it's also probably fine as is - will leave it up to you!